Add ProcessGroupAgent termination detection algorithm

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26984 Add ProcessGroupAgent termination detection algorithm**

closes #26944

In the existing implementation, each worker exits when it sees
no send/recv tasks. However, as we adding support for nested calls,
one RPC could trigger more RPCs in the UDF or in the response
callback. As a result, even if the worker does not see any send/recv
tasks for now, it does not mean there won't be any in the future.

In this commit, we added a counters for all sent and received
messages between each pair of nodes, and then use allgather to collect
those counters, i.e., all workers would have the same view on the
global states. The workers would only exit when all sends are
received and processed.

Differential Revision: [D17633456](https://our.internmc.facebook.com/intern/diff/D17633456)

mrshenli committed:
Add ProcessGroupAgent termination detection algorithm

closes #26944

In the existing implementation, each worker exits when it sees
no send/recv tasks. However, as we adding support for nested calls,
one RPC could trigger more RPCs in the UDF or in the response
callback. As a result, even if the worker does not see any send/recv
tasks for now, it does not mean there won't be any in the future.

In this commit, we added a counters for all sent and received
messages between each pair of nodes, and then use allgather to collect
those counters, i.e., all workers would have the same view on the
global states. The workers would only exit when all sends are
received and processed.

[ghstack-poisoned]

mrshenli committed:
Update on "Add ProcessGroupAgent termination detection algorithm"


closes #26944

In the existing implementation, each worker exits when it sees
no send/recv tasks. However, as we adding support for nested calls,
one RPC could trigger more RPCs in the UDF or in the response
callback. As a result, even if the worker does not see any send/recv
tasks for now, it does not mean there won't be any in the future.

In this commit, we added a counters for all sent and received
messages between each pair of nodes, and then use allgather to collect
those counters, i.e., all workers would have the same view on the
global states. The workers would only exit when all sends are
received and processed.

[ghstack-poisoned]

mrshenli committed:
Update on "Add ProcessGroupAgent termination detection algorithm"


closes #26944

In the existing implementation, each worker exits when it sees
no send/recv tasks. However, as we adding support for nested calls,
one RPC could trigger more RPCs in the UDF or in the response
callback. As a result, even if the worker does not see any send/recv
tasks for now, it does not mean there won't be any in the future.

In this commit, we added a counters for all sent and received
messages between each pair of nodes, and then use allgather to collect
those counters, i.e., all workers would have the same view on the
global states. The workers would only exit when all sends are
received and processed.

Differential Revision: [D17633456](https://our.internmc.facebook.com/intern/diff/D17633456)

[ghstack-poisoned]

pietern commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -159,17 +161,65 @@ void ProcessGroupAgent::join() {
+    // trigger more messages to be sent, we need to wait for the thread pool
+    // again.
+    threadPool_.waitWorkComplete();
+  }
```
`do { .. } while`?

pietern commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -159,17 +161,65 @@ void ProcessGroupAgent::join() {
+    std::lock_guard<std::mutex> guard(sendCntMutex_);
+    std::copy(
+        msgSendCnts_.begin(), msgSendCnts_.end(), cnts.begin() + worldSize);
+  }
```
This can be turned into a simple class `MessageCounter` that deals with addition, the lock, and getting a snapshot (this is all identical between sending and receiving anyway). Then have 2 instances of this thing with an `append` function to add it to the same vector.

Instead of keeping the vector, this could also already be a tensor of size `[2, worldSize]`, to avoid the offsetting math below in the double for loop.

pietern commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -159,17 +161,65 @@ void ProcessGroupAgent::join() {
+
+  for (int i = 0; i < worldSize; ++i) {
+    outputCnts[0].emplace_back(torch::empty({2 * worldSize}, {torch::kInt64}));
+  }
```
We should really have an allgather that does this for you.

pietern commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -275,6 +337,15 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
+          // Not incrementing the recv counter on errors to make it explicit
+          // instead of silent.
+          // TODO: after adding error handling, increment recv counters here
+          // as well.
```
Why? This is a fatal error, right?

pietern commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -64,6 +64,12 @@ class ProcessGroupAgent : public RpcAgent {
+  // recv counters, and then use an allgather to collect all counters from all
+  // peers, i.e., all workers will have the same view of the counters. Then, a
+  // worker will only return true if all send counters match all receive
+  // coutners.
```
Typo in the summary. Also I think this can benefit from a bigger explanation of how the algorithm works.

pietern commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -73,6 +79,15 @@ class ProcessGroupAgent : public RpcAgent {
+  std::vector<int64_t> msgSendCnts_;
+  std::vector<int64_t> msgRecvCnts_;
+  std::mutex sendCntMutex_;
+  std::mutex recvCntMutex_;
```
This would then turn into:

```c++
MessageCounter sendCounts_;
MessageCounter recvCounts_;
```

mrshenli committed:
Update on "Add ProcessGroupAgent termination detection algorithm"


closes #26944

In the existing implementation, each worker exits when it sees
no send/recv tasks. However, as we adding support for nested calls,
one RPC could trigger more RPCs in the UDF or in the response
callback. As a result, even if the worker does not see any send/recv
tasks for now, it does not mean there won't be any in the future.

In this commit, we added a counters for all sent and received
messages between each pair of nodes, and then use allgather to collect
those counters, i.e., all workers would have the same view on the
global states. The workers would only exit when all sends are
received and processed.

Differential Revision: [D17633456](https://our.internmc.facebook.com/intern/diff/D17633456)

[ghstack-poisoned]

pietern commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -159,17 +178,67 @@ void ProcessGroupAgent::join() {
+    // As there could be nested RPC calls, or response callback could also
+    // trigger more messages to be sent, we need to wait for the thread pool
+    // again.
+  } while (!checkNoPendingMessage());
```
The double negation is not super clear here. I think it would be more readable if the method returned the inverse, like `hasPendingMessage()`, so that the condition reads as "continue while there is a pending message", instead of "continue while there is not no pending messages".

pietern commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -159,17 +178,67 @@ void ProcessGroupAgent::join() {
+      // have sendCnt < recv Cnt.
+      if (sentCnt != recvCnt) {
+        return false;
+      }
```
The comment is outdated for the updated check.

pietern reviewed (approved):
Looking good. Two minor comments. Accepting to unblock.

mrshenli committed:
Update on "Add ProcessGroupAgent termination detection algorithm"


closes #26944

In the existing implementation, each worker exits when it sees
no send/recv tasks. However, as we adding support for nested calls,
one RPC could trigger more RPCs in the UDF or in the response
callback. As a result, even if the worker does not see any send/recv
tasks for now, it does not mean there won't be any in the future.

In this commit, we added a counters for all sent and received
messages between each pair of nodes, and then use allgather to collect
those counters, i.e., all workers would have the same view on the
global states. The workers would only exit when all sends are
received and processed.

Differential Revision: [D17633456](https://our.internmc.facebook.com/intern/diff/D17633456)

[ghstack-poisoned]

mrshenli committed:
Update on "Add ProcessGroupAgent termination detection algorithm"


closes #26944

In the existing implementation, each worker exits when it sees
no send/recv tasks. However, as we adding support for nested calls,
one RPC could trigger more RPCs in the UDF or in the response
callback. As a result, even if the worker does not see any send/recv
tasks for now, it does not mean there won't be any in the future.

In this commit, we added a counters for all sent and received
messages between each pair of nodes, and then use allgather to collect
those counters, i.e., all workers would have the same view on the
global states. The workers would only exit when all sends are
received and processed.

Differential Revision: [D17633456](https://our.internmc.facebook.com/intern/diff/D17633456)

[ghstack-poisoned]
